### DISCUSSION TITLE
Logging issue when using tqdm / logging_redirect_tqdm

### DISCUSSION THREAD

**Discussion by fm9000** [apache/airflow#31894] [Q&A]
Hi!

I'm want to have a `PythonOperator` task which is using a package that implements [tqmd](https://github.com/tqdm/tqdm). `tqdm` is used to display pretty progress bars. A bit useless when used within `airflow`, but this package is also used outside of `airflow` in a more interactive environment.

`tqdm` also has a [`logging_redirect_tqdm`](https://tqdm.github.io/docs/contrib.logging/#logging_redirect_tqdm) context manager. It's supposed to capture log messages created within the progressbar-loop and displays it nicely above the progress bar, without having it in a messy interlaced way.

Small example function:

```python
from tqdm import tqdm
from tqdm.contrib.logging import logging_redirect_tqdm
import time

import logging
logger = logging.getLogger("TQDM_Test")

def tqdm_fail(**kwargs):
    i = 10
    pbar = tqdm(total=i)
    with logging_redirect_tqdm():
        for i in range(i):
            logger.error(i)
            time.sleep(0.2)
            pbar.update()
    pbar.close()

tqdm_fail()
```

When running this `tqdm_fail` function from within a `PythonOperator` task, it gets stuck in an endless state, the webinterace becomes freezing/laggy, and the task log is full of entries like:

```
{logging_mixin.py:149} WARNING - --- Logging error ---
{logging_mixin.py:149} WARNING -
{logging_mixin.py:149} WARNING - --- Logging error ---
{logging_mixin.py:149} WARNING - --- Logging error ---
{logging_mixin.py:149} WARNING -
[...]
```

I wanted to open an issue, but the airflow documentation suggests to open a discussion item first. Can you reproduce this problem? I am not sure if the issue is on the `airflow` or `tqdm` side.

**Comment by walter9388** [comment]
I faced the same issue and found that it's because by default `logging_redirect_tqdm()` redirects to the root logger.
Instead, if you pass in `airflow.task` as the logger to `logging_redirect_tqdm()` it should work correctly:
```python
with logging_redirect_tqdm(loggers=[logging.Logger("airflow.task")]):
        ...
```

This led to me adding the following to the simple library I maintain:
```python
# some sort of toggle for whether library is being used in Airflow or not
IS_AIRFLOW = True

# if library is used within airflow, use "airflow.task" for redirect instead of root
loggers: list[logging.Logger] = [logging.root]
if IS_AIRFLOW:
    loggers = [logging.Logger("airflow.task")]

# redirect logging for tqdm
with tqdm_contrib_logging.logging_redirect_tqdm(loggers=loggers):
    ...
```

I still had to disable the progress bars, but the logging works now correctly. For your example above, hopefully something like this should work:
```python
from tqdm import tqdm
from tqdm.contrib.logging import logging_redirect_tqdm
import time

import logging
logger = logging.getLogger("TQDM_Test")

IS_AIRFLOW = True
if IS_AIRFLOW:
    logger = logging.Logger("airflow.task")

def tqdm_fail(**kwargs):
    i = 10
    pbar = tqdm(total=i, disable=IS_AIRFLOW)
    with logging_redirect_tqdm(loggers=[logger]):
        for i in range(i):
            logger.error(i)
            time.sleep(0.2)
            pbar.update()
    pbar.close()

tqdm_fail()
```
